### DISCUSSION TITLE
2.3.0

### DISCUSSION THREAD

**Discussion by mashb1t** [lllyasviel/Fooocus#2559] [General]
## TL;DR
* Add performance "lightning" (based on [SDXL-Lightning 4 step LoRA](https://huggingface.co/ByteDance/SDXL-Lightning/blob/main/sdxl_lightning_4step_lora.safetensors))
* Add preset selection to UI, disable with argument `--disable-preset-selection`. Use `--always-download-new-model` to download missing models on preset switch.
* Add temp path cleanup on startup, configurable in config.txt
* Add support for wildcard subdirectories
* Add scrollable 2 column layout for styles for better structure
* Improve Colab resource needs for T4 instances (default), positively tested with all image prompt features
* Improve anime preset, now uses style `Fooocus Semi Realistic` instead of `Fooocus Negative` (less wet look images)

## What's Changed
* fix: parse width and height as int when applying metadata by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2452
* fix: do not attempt to remove non-existing image grid file by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2456
* feat: add troubleshooting guide to bug report template again by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2489
* feat: use jpeg instead of jpg, use enums instead of strings by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2453
* feat: add performance lightning with 4 step LoRA by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2415
* fix: change synthetic refiner switch from 0.5 to 0.8 by @xhoxye in https://github.com/lllyasviel/Fooocus/pull/2165
* feat: add config for temp path and temp path cleanup on launch by @midareashi and @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1992
* feat: scan wildcard subdirectories by @Cruxial0 and @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2466
* fix: use correct method call for interrupt_current_processing by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2506
* feat: read wildcards in order 通配符增强，切换顺序读取。 by @xhoxye and @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1761
* feat: use scrollable 2 column layout for styles by @hswlab in https://github.com/lllyasviel/Fooocus/pull/1883
* feat: allow to add disabled LoRAs in config by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2507
* fix: prioritize VRAM over RAM in Colab, preventing out of memory issues by @Peppe289 and @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1710
* fix: update xformers to 0.0.23 by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2517
* feat: update xformers to 0.0.23 in Dockerfile by @josephrocca in https://github.com/lllyasviel/Fooocus/pull/2519
* fix: parse seed as string to display correctly in metadata preview by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/2536
* feat: allow user add custom preset without block `git pull` by @Zxilly in https://github.com/lllyasviel/Fooocus/pull/2520
* feat: add preset selection to Gradio UI (session based) by @mashb1t in https://github.com/lllyasviel/Fooocus/pull/1570
* fix: add error output for unsupported images by @shaye059 in https://github.com/lllyasviel/Fooocus/pull/2537
* feat: improve anime preset by adding style Fooocus Semi Realistic by @DavidDragonsage in https://github.com/lllyasviel/Fooocus/pull/2492
* feat: add backwards compatibility for presets without disable/enable LoRA boolean by @mashb1t

## New Contributors
* @midareashi made their first contribution in https://github.com/lllyasviel/Fooocus/pull/1992
* @Cruxial0 made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2466
* @hswlab made their first contribution in https://github.com/lllyasviel/Fooocus/pull/1883
* @Peppe289 made their first contribution in https://github.com/lllyasviel/Fooocus/pull/1710
* @josephrocca made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2519
* @Zxilly made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2520
* @shaye059 made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2537
* @DavidDragonsage made their first contribution in https://github.com/lllyasviel/Fooocus/pull/2492

**Full Changelog**: https://github.com/lllyasviel/Fooocus/compare/2.2.1...2.3.0

<hr /><em>This discussion was created from the release <a href='https://github.com/lllyasviel/Fooocus/releases/tag/2.3.0'>2.3.0</a>.</em>

**Comment by CadLoop** [comment]
As always infinite gratitude for this great work! :) Whenever I go, I start conversation "do you know this epic github repo?"

**Comment by Tra2x** [comment]
With reach release update I have some Issues after the automatic update. Any Ideas? 
Most probably it has to do with my automatic replace of the config.txt (I use different ones for normal image creation when my kids are also with me and the 'private' images ;) )

> Fooocus startup
traxx@traxx-ubuntu:~$ cat ./x_fooocus.sh 
#!/bin/bash
cd Fooocus
rm config.txt
cp config_x.txt config.txt
source fooocus_env/bin/activate
export HSA_OVERRIDE_GFX_VERSION=10.3.0
export HCC_AMDGPU_TARGET=gfx1032
python entry_with_update.py --always-low-vram


> Fooocus Log
Already up-to-date
 Update succeeded.   
[System ARGV] ['entry_with_update.py', '--always-low-vram']
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
Fooocus version: 2.3.0
Failed to load config key: {"default_aspect_ratio": "1152*896"} is invalid; will use {"default_aspect_ratio": "1024*1024"} instead.
[Cleanup] Attempting to delete content of temp dir /tmp/fooocus
[Cleanup] Cleanup successful
Total VRAM 8176 MB, total RAM 48091 MB
Set vram state to: LOW_VRAM
Always offload VRAM
Device: cuda:0 AMD Radeon RX 6650 XT : native
VAE dtype: torch.float32
Using sub quadratic optimization for cross attention, if you have memory or speed issues try using: --attention-split
Exception in thread Thread-2 (worker):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/home/traxx/Fooocus/modules/async_worker.py", line 32, in worker
    import modules.default_pipeline as pipeline
  File "/home/traxx/Fooocus/modules/default_pipeline.py", line 257, in <module>
Traceback (most recent call last):
  File "/home/traxx/Fooocus/entry_with_update.py", line 46, in <module>
    from launch import *
  File "/home/traxx/Fooocus/launch.py", line 136, in <module>
    from webui import *
  File "/home/traxx/Fooocus/webui.py", line 364, in <module>
    for i, (enabled, filename, weight) in enumerate(modules.config.default_loras):
ValueError: not enough values to unpack (expected 3, got 2)
    loras=get_enabled_loras(modules.config.default_loras)
  File "/home/traxx/Fooocus/modules/util.py", line 366, in get_enabled_loras
    return [[lora[1], lora[2]] for lora in loras if lora[0]]
  File "/home/traxx/Fooocus/modules/util.py", line 366, in <listcomp>
    return [[lora[1], lora[2]] for lora in loras if lora[0]]
IndexError: list index out of range


Thanks in advance

**Reply by cuffins** [reply]
It seems there is a general issue with the "config.txt". With every start Fooocus is writing a new "config_modification_tutorial.txt" and ignores the entries in "config.txt". Maybe this is the reason for your problem.

**Reply by Tra2x** [reply]
Yes, I used the original config.txt and replaced my values. Now it's working fine. Thanks

**Reply by mashb1t** [reply]
Can you please provide the previous config.txt leading to the error mentioned above?

**Reply by Tra2x** [reply]
[config_x.txt](https://github.com/lllyasviel/Fooocus/files/14656937/config_x.txt)
This is the working file. When I add `"default_max_lora_number": 10,` then I get the mentioned error.

Thanks for asking (y)

**Reply by mashb1t** [reply]
fixed in https://github.com/lllyasviel/Fooocus/pull/2576

**Comment by bsdesigner** [comment]
What should I write in the config.txt file to have these checkboxes enabled by default?
![Снимок экрана 2024-03-19 190043](https://github.com/lllyasviel/Fooocus/assets/36730003/a6505c8e-65a2-4ba6-a140-562a0c9591fd)

**Reply by mashb1t** [reply]
You'll have to change the code in https://github.com/lllyasviel/Fooocus/blob/main/webui.py#L412 to True as this parameter isn't configurable.

**Reply by bsdesigner** [reply]
Thanks for the quick reply! But I'm a complete dummy about codes, so I'd better stay out of it ))

**Reply by bsdesigner** [reply]
Figured it out, did it, it worked, it's all on. Thanks again for the tip!

**Reply by mashb1t** [reply]
Just be aware of that the automated update might not work anymore when doing changes in your local environment

**Comment by generativekey** [comment]
I've used all the UIs you can imagine, but nothing compares to this one. The ease, consistency, speed, practicality, and comfort that this UI offers is out of this world. You're truly doing an amazing job. I believe that Fooocus is the best UI we have today.

Not to forget mentioning that the upscale and inpainting really work wonders! Very well done! I am very grateful, friends.

**Comment by Kira8822** [comment]
Will we have this version 2.3.0 to work offline locally for Windows?

**Reply by mashb1t** [reply]
@Kira8822 Fooocus always works offline if you already have downloaded all the necessary models. If thr background of your question is referring to a new zip file then the answer is no, this will only be created for python package updates, which wasn't the case in 2.3.0.

**Comment by Kira8822** [comment]
Excuse my bad English, I'm with the translator, yes that's what I was referring to a new zip file, because version 2.3.0 is really very good, especially because you prioritized the load it had in RAM over the GPU, instead in the zip file doesn't happen so it consumes a lot of ram and that's what is happening to me, even though I have 32 GB of ram it consumes a lot and that GPU prioritization section in the zip would be nice. I don't know if you could implement it if it's not too much trouble?

**Reply by mashb1t** [reply]
@Kira8822 The Fooocus version in the zip file will automatically update to the latest one on first start, no need to create a new zip file. If you need to have an "offline" zip file, feel free to zip your current folder after creating images, downloading all necessary models and loras etc.
RAM and GPU usage has been optimized in Colab, you can also use the same parameters locally by adding them to your run.bat.
`--always-high-vram` is the argument which was added in Colab, but there are more possible optimisations you can find in the mentioned PR above.

**Reply by Kira8822** [reply]
I understand, thank you...

**Comment by Cochonfou** [comment]
Hello, and thanks a lot for the update.
I am using Fooocus on a M3 Pro mac. The new "lightning" performance setting does not seem to work on my configuration, as it always results in black images or dark color blotches. Everything else works fine as far as I know.

Would you have an idea of what could be wrong ?

Thanks a lot.

![image(2)](https://github.com/lllyasviel/Fooocus/assets/158749893/57ab9868-9db9-48a1-aa93-e7b59f0abb63)

```
{
"prompt": "steam train",
"negative_prompt": "",
"prompt_expansion": "steam train, detailed, elegant, highly color saturated, intricate, full perfect, sharp focus, beautiful, epic composition, cinematic, amazing light, dynamic background, advanced, atmosphere, lively, magical, very inspirational, professional, decorated, stunning, inspired, creative, positive, trendy, cute, adorable, pretty, pure, coherent, fine detail, polished, complex, enhanced",
"styles": "['Fooocus V2', 'Fooocus Enhance', 'Fooocus Sharp']",
"performance": "Lightning",
"resolution": "(1152, 896)",
"guidance_scale": 1.0,
"sharpness": 0.0,
"adm_guidance": "(1.0, 1.0, 0.0)",
"base_model": "juggernautXL_v8Rundiffusion.safetensors",
"refiner_model": "None",
"refiner_switch": 1.0,
"adaptive_cfg": 1.0,
"sampler": "euler",
"scheduler": "sgm_uniform",
"seed": "5000134209646226646",
"lora_combined_6": "sdxl_lightning_4step_lora.safetensors : 1.0",
"metadata_scheme": false,
"version": "Fooocus v2.3.0"
}
```

```
Already up-to-date
Update succeeded.
[System ARGV] ['entry_with_update.py']
Python 3.10.14 (main, Mar 21 2024, 11:21:31) [Clang 14.0.6 ]
Fooocus version: 2.3.0
[Cleanup] Attempting to delete content of temp dir /var/folders/m4/w360gldx1xx7nb2v895_533w0000gn/T/fooocus
[Cleanup] Cleanup successful
Total VRAM 36864 MB, total RAM 36864 MB
Set vram state to: SHARED
Always offload VRAM
Device: mps
VAE dtype: torch.float32
Using sub quadratic optimization for cross attention, if you have memory or speed issues try using: --attention-split
Refiner unloaded.
Running on local URL:  http://127.0.0.1:7865

To create a public link, set `share=True` in `launch()`.
model_type EPS
UNet ADM Dimension 2816
Using split attention in VAE
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
Using split attention in VAE
extra {'cond_stage_model.clip_l.text_projection', 'cond_stage_model.clip_l.logit_scale', 'cond_stage_model.clip_g.transformer.text_model.embeddings.position_ids'}
Base model loaded: /Users/Shared/Fooocus/models/checkpoints/juggernautXL_v8Rundiffusion.safetensors
Request to load LoRAs [['sd_xl_offset_example-lora_1.0.safetensors', 0.1], ['None', 1.0], ['None', 1.0], ['None', 1.0], ['None', 1.0]] for model [/Users/Shared/Fooocus/models/checkpoints/juggernautXL_v8Rundiffusion.safetensors].
Loaded LoRA [/Users/Shared/Fooocus/models/loras/sd_xl_offset_example-lora_1.0.safetensors] for UNet [/Users/Shared/Fooocus/models/checkpoints/juggernautXL_v8Rundiffusion.safetensors] with 788 keys at weight 0.1.
Fooocus V2 Expansion: Vocab with 642 words.
Fooocus Expansion engine loaded for cpu, use_fp16 = False.
Requested to load SDXLClipModel
Requested to load GPT2LMHeadModel
Loading 2 new models
Started worker with PID 37397
App started successful. Use the app with http://127.0.0.1:7865/ or 127.0.0.1:7865
Loaded preset: /Users/Shared/Fooocus/presets/lightning.json
Enter Lightning mode.
[Fooocus] Downloading Lightning components ...
[Parameters] Adaptive CFG = 1.0
[Parameters] Sharpness = 0.0
[Parameters] ControlNet Softness = 0.25
[Parameters] ADM Scale = 1.0 : 1.0 : 0.0
[Parameters] CFG = 1.0
[Parameters] Seed = 5000134209646226646
[Parameters] Sampler = euler - sgm_uniform
[Parameters] Steps = 4 - 4
[Fooocus] Initializing ...
[Fooocus] Loading models ...
Refiner unloaded.
Request to load LoRAs [['None', 1.0], ['None', 1.0], ['None', 1.0], ['None', 1.0], ['None', 1.0], ('sdxl_lightning_4step_lora.safetensors', 1.0)] for model [/Users/Shared/Fooocus/models/checkpoints/juggernautXL_v8Rundiffusion.safetensors].
Loaded LoRA [/Users/Shared/Fooocus/models/loras/sdxl_lightning_4step_lora.safetensors] for UNet [/Users/Shared/Fooocus/models/checkpoints/juggernautXL_v8Rundiffusion.safetensors] with 788 keys at weight 1.0.
Requested to load SDXLClipModel
Loading 1 new model
unload clone 1
[Fooocus] Processing prompts ...
[Fooocus] Preparing Fooocus text #1 ...
[Prompt Expansion] steam train, detailed, elegant, highly color saturated, intricate, full perfect, sharp focus, beautiful, epic composition, cinematic, amazing light, dynamic background, advanced, atmosphere, lively, magical, very inspirational, professional, decorated, stunning, inspired, creative, positive, trendy, cute, adorable, pretty, pure, coherent, fine detail, polished, complex, enhanced
[Fooocus] Encoding positive #1 ...
[Parameters] Denoising Strength = 1.0
[Parameters] Initial Latent shape: Image Space (896, 1152)
Preparation time: 2.18 seconds
[Sampler] refiner_swap_method = joint
[Sampler] sigma_min = 0.6951494812965393, sigma_max = 14.614640235900879
Requested to load SDXL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 9.85 seconds
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
100%|█████████████████████████████████████████████| 4/4 [00:11<00:00,  2.99s/it]
Requested to load AutoencoderKL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 1.13 seconds
Image generated with private log at: /Users/Shared/Fooocus/outputs/2024-03-23/log.html
Generating and saving time: 25.87 seconds
Total time: 28.08 seconds
```

**Reply by mashb1t** [reply]
=> https://github.com/lllyasviel/Fooocus/issues/2580
Can't reproduce with my M1 Macbook Pro, works for me.

**Reply by Cochonfou** [reply]
Thanks I will continue in the bug ticket.
